docs/design.md

```diff
@@ -1,87 +1,133 @@
+- Note that Kubernetes may time out waiting for ControllerUnpublishVolume too. In this case, it will keep calling it until it gets confirmation from the driver that the volume has been detached (i.e. until ebs-plugin returns either success or non-timeout error) or it needs the volume attached to the node again (and it will call ControllerPublishVolume in that case).
 - The same applies to NodeStage and NodePublish calls (&quot;mount device, mount volume&quot;). These are typically much faster than attach/detach, still they must be idempotent when it comes to timeouts.
 
-It looks complicated, but it should be actually simple - always check that if the required operation has been already done
+In summary, always check that if the required operation has already been done.
 
 ### Restarts
 
 The CSI driver should survive its own crashes or reboots of the node where it runs. For the controller service, Kubernetes will either start a new driver on a different node or re-elect a new leader of stand-by drivers. For the node service, Kubernetes will start a new driver shortly.
 
-The perfect CSI driver should be stateless. After start, it should recover its state by observing the actual status of AWS (i.e. describe instances / volumes). Current cloud provider follows this approach, however there are some corner cases around restarts when Kubernetes can try to attach two volumes to the same device on a node.
-
-When the stateless driver is not possible, it can use some persistent storage outside of the driver. Since the driver should support multiple Container Orchestrators (like Mesos), it must not use Kubernetes APIs. It should use AWS APIs instead to persist its state if needed (like AWS DynamoDB). We assume that costs of using such db will be negligible compared to rest of Kubernetes.
+The ideal CSI driver is stateless. After start, it should recover its state by observing the actual status of AWS (i.e. describe instances / volumes).
 
 ### No credentials on nodes
 
 General security requirements we follow in Kubernetes is &quot;if a node gets compromised then the damage is limited to the node&quot;. Paranoid people typically dedicate handful of nodes in Kubernetes cluster as &quot;infrastructure nodes&quot; and dedicate these nodes to run &quot;infrastructure pods&quot; only. Regular users can&#39;t run their pods there. CSI attacher and provisioner is an example of such &quot;infrastructure pod&quot; - it need permission to create/delete any PV in Kubernetes and CSI driver running there needs credentials to create/delete volumes in AWS.
 
 There should be a way how to run the CSI driver (=container) in &quot;node mode&quot; only. Such driver would then respond only to node service RPCs and it would not have any credentials to AWS (or very limited credentials, e.g. only to Describe things). Paranoid people would deploy CSI driver in &quot;node only&quot; mode on all nodes where Kubernetes runs user containers.
```

Comment:
Thoughts on also removing the "No credentials on nodes" section? (which I understand has been in here for several years). It adds an arbitrary restriction, its not actually a design requirement. At the very least,  please replace "Paranoid people" with "security conscious users".

Reply:
Hmm, I'm inclined to keep this section because I believe it lays out a few great points on why node mode container should have limited credentials.

Will replace 'paranoid' wording though. Thanks!

Reply:
I am still a bit confused about the second paragraph of this chapter. Why should people run an EBS CSI driver in "node only" mode which is not able to provision any EBS volume (because it only has Describe* permissions)

Reply:
Both the EBS CSI Driver Controller Deployment and Node Daemonset pods have the same `ebs-plugin` container. The difference is that in the controller pod we set the plugin's mode to `controller`, and in the node pods we set the plugin's mode to `node`. For example, this node mode is set on Daemonset's ebs-plugin container in the helm template [here](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/39822b5ceb4d8565e15039715dfdc622c14fa6e5/charts/aws-ebs-csi-driver/templates/_node.tpl#L72)

Because the node pods set the ebs-plugin container to node mode, it won't respond to Controller Service RPCs (ControllerCreateVolume). 

Today, we also don't give any of the daemonset's pods access to the EBS CSI Controller IAM Role, because the Node Service RPCs like NodeStageVolume don't need to talk to the EC2 API. 

Finally, we ensure only the EBS CSI Controller pods have the higher-risk K8s RBAC for actions like patching PV and VolumeAttachment  resources. 

Because of the 3 points above, a security conscious customer can ensure that the EBS CSI Controller pods are only scheduled on extra hardened nodes. If an intruder gains access to any of the other nodes on the cluster they wouldn't be able to create/delete EBS volumes via the `ebs-csi-node` pod, or mess with the cluster's storage resources. 

Did I understand your question correctly @youwalther65?

Reply:
Exactly. Great, I'd like to see this answer in the doc itself!